core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java

```diff
@@ -58,8 +58,17 @@ private static final class WeightedRoundRobinSelector extends AbstractEndpointSe
 
         WeightedRoundRobinSelector(EndpointGroup endpointGroup) {
             super(endpointGroup);
-            endpointGroup.addListener(endpoints -> endpointsAndWeights = new EndpointsAndWeights(endpoints),
-                                      true);
+            endpointGroup.whenReady().thenAccept(unused -> maybeUpdateEndpoints(endpointGroup.endpoints()));
```

Comment:
Minor question:
Was this also an optimization? or is this change necessary? (It's hard to track which changes correspond to which motivation 😅 )

I prefer that we always try to use listeners instead of `whenReady` since:
1) Users usually call `whenReady.join()` on application startup. If listeners are invoked beforehand there is less probability of race conditions
2) Less complexity - we only have to care about a single callback. I feel like most of this confusion is because we are mixing usages of `initialCompleteFuture` with `listener`s

```suggestion
            endpointGroup.addListener(endpoints -> this.endpointsAndWeights = new EndpointsAndWeights(endpoints));
```

Reply:
I removed the whole code but let me explain my original purpose. I changed `AbstractEndpointSelector()` to add an endpoint listener first when it is created which resulted in adding the listener of `WeightedRoundRobinSelector` later. Now we have two endpoint listeners.

If the `EndpointGroup` is not initialized yet, `AbstractEndpointSelector.select()` creates `ListeningFuture` and waits for new endpoints via the listener. When `AbstractEndpointSelector` receives the event, it consequently calls `WeightedRoundRobinSelector.selectNow()`. At the moment, the listener of `WeightedRoundRobinSelector` is not invoked yet and its `SelectionSterategy` will be empty or null.

For that reason, I used `whenReady()` to initialize `SelectionSterategy` before `AbstractEndpointSelector` receives an endpoint update event.